### PR TITLE
Blue Billywig bid adapter

### PR DESCRIPTION
**Pull request by K-JBoon**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change
Adds the Blue Billywig bid adapter, sample ad unit in bluebillywigBidAdapter.md

- test parameters for validating bids (outstream video!)
```
{
  bidder: 'bluebillywig',
  params: {
    publicationName: 'bbprebid',
    accountId: 642
    rendererCode: 'renderer',
    connections: ['bluebillywig'],
    bluebillywig: {}
  }
}
```
- contact email of the adapter’s maintainer: dev+prebid@bluebillywig.com
- [x] official adapter submission

## Other information

prebid.github.io PR: https://github.com/prebid/prebid.github.io/pull/1921

### CODE REVIEW

**Comment by K-JBoon** [comment]
@mike-chowla Hey Mike, could you give some insight on why the CircleCI build is failing? I can't check the details of that report

**Comment by mike-chowla** [comment]
@K-JBoon CircleCI log:
[bbw.txt](https://github.com/prebid/Prebid.js/files/4507323/bbw.txt)

File: modules/bluebillywigBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+    if (adServerCur && typeof adServerCur === 'string') request.cur = [adServerCur];
+    else if (Array.isArray(adServerCur) && adServerCur.length) request.cur = [adServerCur[0]];
+  },
+  addUserIds: (request, validBidRequests) => {
+    if (!request) return;
+
+    // NB straight rip from prebidServerAdapter, keep track for updates
+    const bidUserId = utils.deepAccess(validBidRequests, '0.userId');
+
+    if (bidUserId && typeof bidUserId === 'object' && (bidUserId.tdid || bidUserId.pubcid || bidUserId.parrableid || bidUserId.lipb || bidUserId.id5id || bidUserId.criteoId || bidUserId.britepoolid || bidUserId.idl_env)) {
```
Comment (mike-chowla) [inline modules/bluebillywigBidAdapter.js:63]: Instead constructing eids one by one, you can just use this function:
https://github.com/prebid/Prebid.js/blob/c0c8b58e4c8622c501fab83a8dd6ef3fe4fa3eec/modules/userId/eids.js#L120

File: modules/bluebillywigBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+import { registerBidder } from '../src/adapters/bidderFactory.js';
+import adapterManager from '../src/adapterManager.js';
+import { VIDEO } from '../src/mediaTypes.js';
+import { config } from '../src/config.js';
+import { Renderer } from '../src/Renderer.js';
+
+const DEV_MODE = window.location.search.match(/bbpbs_debug=true/);
+
+// Blue Billywig Constants
+export const BB_CONSTANTS = {
```
[the remark this answers is not in the document]
Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.js:11]: The constants are currently being used in the test spec to check that default values are as defined in the original file, and to supply the bidder code. I suppose the idiomatic way is to copy these into the test file and update in both places when needed?

File: modules/bluebillywigBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+  DEFAULT_WIDTH: 768,
+  DEFAULT_HEIGHT: 432,
+  DEFAULT_NET_REVENUE: true
+};
+
+// Aliasing
+const getConfig = config.getConfig;
+
+// Helper Functions
+export const BB_HELPERS = {
```
[the remark this answers is not in the document]
Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.js:27]: Previously used in the test file, decided that didn't make sense, forgot to remove it there. Will fix.

File: modules/bluebillywigBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+  getSyncUrl: (publication) => {
+    return BB_HELPERS.substituteUrl(BB_CONSTANTS.SYNC_URL, publication);
+  },
+  getRendererUrl: (publication, renderer) => {
+    return BB_HELPERS.substituteUrl(BB_CONSTANTS.RENDERER_URL, publication, renderer);
+  },
+  getAliasesFromRegistry: (name) => {
+    if (!name) return null;
+
+    let aliases = adapterManager.aliasRegistry[name];
```
[the remark this answers is not in the document]
Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.js:177]: Pattern taken from the s2s adapter which was partially used as a base, shouldn't actually be needed here. Will remove.

File: modules/bluebillywigBidAdapter.js (hunk 1 above)
Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.js:63]: Didn't know, cheers! Will fix.

File: modules/bluebillywigBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+
+    if (aliases) return aliases;
+    else return null;
+  },
+  getBidAdapterFromManager: (name) => {
+    if (!name) return null;
+
+    const adapter = adapterManager.getBidAdapter && adapterManager.getBidAdapter(name);
```
[the remark this answers is not in the document]
Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.js:185]: Same as aliases, copied from S2S adapter which was used as a base, shouldn't actually be needed here. Will fix.

**Comment by K-JBoon** [comment]
Hey @mike-chowla , thanks for the review! I've updated the code according to your recommendations/comments.

As an aside, is there a style guide/spec for the actual adapter code? CircleCI fell over using `for .. of` loops, but can't find what I can/can't use (given that the build process does succeed locally, and the practices across adapters vary a lot).

File: modules/bluebillywigBidAdapter.js (hunk 1 above)
Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.js]: Rewrote to use createEidsArray

File: modules/bluebillywigBidAdapter.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.js]: Removed.

File: modules/bluebillywigBidAdapter.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.js]: Removed.

File: modules/bluebillywigBidAdapter.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.js:27]: Removed.

File: modules/bluebillywigBidAdapter.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.js]: Removed.

**Comment by K-JBoon** [comment]
Hey @mike-chowla , any update on this?

**Comment by mike-chowla** [comment]
@K-JBoon 

The CircleCI output shows all the failures on IE 11, I'm guessing for..of loops are not supported on IE 11

`IE 11.0.0 (Windows 10.0.0): Executed 5485 of 5488 (7 FAILED) (skipped 3) (17.122 secs / 9.997 secs)`

Can you fix both the CircleCI errors and the LGTM one?

**Comment by K-JBoon** [comment]
Hey @mike-chowla , I removed the "for ... of" in the recent commit. I can't see the CircleCI log output even when authorizing it, I guess because I'm not part of this repository's contributors? Could you send that log over again or give some directions to getting access to it?

I'll fix the LGTM alert. Cheers!

**Comment by K-JBoon** [comment]
@mike-chowla Removed a `for ... of` that I missed, and replaced all the lambda arrows with regular function, and removed use of `Array.prototype.find`

LGTM passed now but didn't see another CircleCI run

**Comment by K-JBoon** [comment]
Hey @mike-chowla , any chance you could check if it can do a CircleCI run or if it just passes now? Cheers!

**Comment by mike-chowla** [comment]
I ran the Browserstack tests and you still have one failure on IE 11:

> Chrome 79.0.3945 (Windows 10.0.0): Executed 5551 of 5554 (skipped 3) SUCCESS (10.26 secs / 5.389 secs)
> Chrome 80.0.3987 (Windows 10.0.0): Executed 5551 of 5554 (skipped 3) SUCCESS (10.447 secs / 5.416 secs)
> Edge 18.18363.0 (Windows 10.0.0): Executed 5551 of 5554 (skipped 3) SUCCESS (14.804 secs / 8.774 secs)
> IE 11.0.0 (Windows 10.0.0): Executed 5551 of 5554 (1 FAILED) (skipped 3) (13.944 secs / 8.821 secs)
> Firefox 73.0.0 (Windows 10.0.0): Executed 5551 of 5554 (skipped 3) SUCCESS (12.621 secs / 6.34 secs)
> Firefox 72.0.0 (Windows 10.0.0): Executed 5551 of 5554 (skipped 3) SUCCESS (13.678 secs / 6.47 secs)
> Safari 13.0.2 (Mac OS X 10.15.0): Executed 5551 of 5554 (skipped 3) SUCCESS (9.219 secs / 5.876 secs)
> Safari 12.1.2 (Mac OS X 10.14.6): Executed 5551 of 5554 (skipped 3) SUCCESS (11.459 secs / 7.44 secs)
> Edge 17.17134.0 (Windows 10.0.0): Executed 5551 of 5554 (skipped 3) SUCCESS (15.28 secs / 8.32 secs)
> TOTAL: 1 FAILED, 49958 SUCCESS
> 
> 
> 1) should append the various values if they exist
>      BlueBillywigAdapter getUserSyncs
>      ReferenceError: 'Symbol' is undefined
>    at getUserSyncs (webpack:///modules/bluebillywigBidAdapter.js:351:5 <- test/test_index.js:108250:12)
>    at Anonymous function (webpack:///test/spec/modules/bluebillywigBidAdapter_spec.js:887:7 <- test/test_index.js:107865:7)

**Comment by K-JBoon** [comment]
@mike-chowla Cheers! I swear I'm blind as a bat... That should be the last one, sorry about that.

File: modules/bluebillywigBidAdapter.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+                        context: 'outstream',
+                        mimes: ['video/mp4'],
+			protocols: [ 2,3,5,6]
+                    }
+                },
+                bids: [{
+                    bidder: 'bluebillywig',
+                    params: {
+			publicationName: "bbprebid",
+			renderer: "renderer",
```
Comment (mike-chowla) [inline modules/bluebillywigBidAdapter.md:31]: These test parameters get rejected as invalid due to lack of rendererCode.  When I change, renderer to rendererCode, I get an ad returned

Comment (K-JBoon) [reply modules/bluebillywigBidAdapter.md]: Oops, yeah, forgot that one when refactoring the name. Fixed!

**Comment by K-JBoon** [comment]
@mike-chowla Fixed now, cheers!

**Comment by mike-chowla** [comment]
Docs PR is https://github.com/prebid/prebid.github.io/pull/1921

**Pull request merged**